### ISSUE TITLE
enumerate() segfaults on Windows when running a x86 binary on a x64 system

### ISSUE THREAD

**Issue by ned14** [BoostGSoC13/boost.afio#79]
This is due to a bug in the WOW64 syscall translation layer. Microsoft have told me they won't fix it as I'm not supposed to be using NT kernel APIs directly, so I'll need to figure out a workaround.

**Comment by ariccio** [comment]
*Maybe* `AFIO` could fall back (for `x86` builds) to an async *dispatched* version of `FindFirstFile` (and friends)?

**Comment by ariccio** [comment]
Referencing: http://ariccio.com/2015/01/13/how-fast-are-findfirstfilefindfirstfileex-and-cfilefind-actually/


-----

*The most naive* implementation of async dispatching (using `std::async`) is still considerably faster than *no* asynchrony. In that implementation is spent spinning in `get` (the async event destructor). If you use `AFIO`'s `IOCP`/Closure Execution Engine framework, it'll probably be much faster still.

**Comment by ned14** [comment]
I'll be rewriting the engine anyway for v1.4 after the Boost peer review. This is a reminder for me to implement a workaround.

**Comment by ariccio** [comment]
Ah, ok. That'll be interesting... What sort of design changes are you thinking of?

**Comment by ned14** [comment]
I have an outline of the changes for 1.4 at https://boostgsoc13.github.io/boost.afio/doc/html/afio/design.html.

I have the two year plan at the top of https://boostgsoc13.github.io/boost.afio/doc/html/afio/release_notes.html.

Are these changes going to break any code for you?

**Comment by ariccio** [comment]
> Are these changes going to break any code for you?

Nah, I haven't (*yet*) written any real code that uses `AFIO`. I'm very carefully following the development of `AFIO` for several reasons (it's interesting, I plan to use it, etc...), so I'm ok here.



I'm wondering, as `AFIO` ***is*** using the undocumented/unsupported NT Native API, maybe it'd be a good idea to have support for that async dispatched version of `FindFirstFile`, as a fallback mechanism, if for some reason `AFIO` can't use the kernel API? Hell, it could even be an automatic fallback, so user (*i.e. library-user*) code doesn't need to explicitly ask for it.

**Comment by ned14** [comment]
You may have noticed AFIO hasn't begun peer review yet. This is because I am unhappy with presenting the library in its current form, so I've asked for advice. I may end up simply rewriting the tutorial to match the new API propoerly, or push out the review for six months till after the lightweight future promise refactor.

> I'm wondering, as AFIO is using the undocumented/unsupported NT Native API, maybe it'd be a good idea to have support for that async dispatched version of FindFirstFile, as a fallback mechanism, if for some reason AFIO can't use the kernel API? Hell, it could even be an automatic fallback, so user (i.e. library-user) code doesn't need to explicitly ask for it.

I'm not at all concerned about using the NT kernel API directly. It's amazingly stable over time, they almost never change it. The big reasons for avoiding the win32 layer is it's very slow, and it's full of quirks. You definitely don't want to mix paradigms like AFIO < 1.3 did, it's a never ending source of hard to find bugs.

Regarding fixing this here issue, the bug in WOW64 is only an issue here because I can't patch ASIO for it. If I could patch ASIO, the problem of the IOCP pointer getting randomly 32 bit sliced goes away.

In my planned rewrite of the engine around lightweight futures, the workaround for 32 bit on 64 bit is silly trivial - don't hand off enumeration completion handling to ASIO. That's why I didn't bother with fixing this now as it's all wasted effort after lightweight futures.

Lightweight futures + coroutines may even replace ASIO if WG21 decides to allow coroutines to auto schedule across threads. Just start N threads and have them coroutine await, the runtime automatically fills the threads with coroutine work for you. What Gor proposed can do this, but nobody is currently sure if that will enter the final TS. If so, then our need for an ASIO dependency is just the scatter gather buffers, and I can drop ASIO/the Networking TS completely if on a compiler with C++ 1z coroutines.

**Comment by ariccio** [comment]
> In my planned rewrite of the engine around lightweight futures, the workaround for 32 bit on 64 bit is silly trivial - don't hand off enumeration completion handling to ASIO. That's why I didn't bother with fixing this now as it's all wasted effort after lightweight futures.

Ahh, this sounds very cool - I think I may need to go back and read over the whole "lightweight futures" thing. Would that be *as lightweight* as IOCP?

**Comment by ned14** [comment]
> Would that be as lightweight as IOCP?

It's hard to say. I haven't played with Microsoft's coroutines engine yet, so until I do I cannot say.

I would assume, however, that MSVC is as constrained by OS facilities as ASIO is.
